### PR TITLE
Fix 16092: Fix duplicate joins in usage of joinWith

### PR DESCRIPTION
**Pull request by germanow**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #16092

### CODE REVIEW

File: tests/framework/db/ActiveQueryTest.php [hunk 1]
```diff
@@ -136,6 +136,26 @@ public function testInnerJoinWith()
         ], $result->joinWith);
     }
 
+    public function testBuildJoinWithRemoveDuplicateJoinByTableName()
+    {
+        $query = new ActiveQuery(Customer::className());
+        $query->innerJoinWith('orders')
+            ->joinWith('orders.orderItems');
```
Comment (xepozz) [inline tests/framework/db/ActiveQueryTest.php]: What about closure as a `on` statement?
https://www.yiiframework.com/doc/guide/2.0/en/db-active-record#relation-table-aliases

**Review by xepozz — CHANGES_REQUESTED** [review]

File: tests/framework/db/ActiveQueryTest.php (hunk 1 above)
Comment (germanow) [reply tests/framework/db/ActiveQueryTest.php]: Do you mean this case?
```php
$query = new ActiveQuery(Customer::className());
$query->innerJoinWith('orders')
    ->joinWith('orders', function($q) {
        $q->joinWith('orderItems');
    });
```
Works well.

Comment (xepozz) [reply tests/framework/db/ActiveQueryTest.php]:
```php
$query->joinWith([
    'orders' => function ($q) {
        $q->from(['o' => Order::tableName()]);
    },
])
```
and
```php
$query->joinWith([
    'orders' => function ($q) use ($value) {
        $q->where(['key' => $value]);
    },
])
```

Comment (germanow) [reply tests/framework/db/ActiveQueryTest.php]: I added tests for it to ActiveRecordTest. Why build failed after that? Failure in `FixtureControllerTest`: 
```
1) yiiunit\framework\console\controllers\FixtureControllerTest::testLoadActiveFixtureSequence
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => 'yiiunit\data\console\controllers\fixtures\SecondIndependentActiveFixture'
-    1 => 'yiiunit\data\console\controllers\fixtures\FirstIndependentActiveFixture'
+    0 => 'yiiunit\data\console\controllers\fixtures\FirstIndependentActiveFixture'
+    1 => 'yiiunit\data\console\controllers\fixtures\SecondIndependentActiveFixture'
     2 => 'yiiunit\data\console\controll...ixture'
 )

/home/runner/work/yii2/yii2/tests/framework/console/controllers/FixtureControllerTest.php:244
/home/runner/work/yii2/yii2/vendor/phpunit/phpunit/phpunit:52
```

Comment (samdark) [reply tests/framework/db/ActiveQueryTest.php]: That test was incorrect. I've fixed it.

File: tests/framework/db/ActiveRecordTest.php [hunk 2]
```diff
@@ -561,6 +561,134 @@ public function testJoinWith()
         $this->assertEquals(2, $orders[0]->items[0]->category->id);
     }
 
+    /**
+     * @depends testJoinWith
+     */
+    public function testJoinWithDuplicate()
```
Comment (xepozz) [inline tests/framework/db/ActiveRecordTest.php]: Could you please use `@dataProvider` if it's possible or split this long method to small methods if not?

File: framework/db/ActiveQuery.php [hunk 3]
```diff
@@ -486,6 +486,16 @@ private function buildJoinWith()
         }
         $this->join = array_values($uniqueJoins);
 
+        // https://github.com/yiisoft/yii2/issues/16092
+        $uniqueJoinsByTableName = [];
+        for ($i = 0; $i < count($this->join); $i++) {
```
Comment (xepozz) [inline framework/db/ActiveQuery.php]: I think here `foreach` will be best look. What do you think?

**Review by xepozz — CHANGES_REQUESTED** [review]
Everything looks good

File: tests/framework/db/ActiveRecordTest.php (hunk 2 above)
Comment (germanow) [reply tests/framework/db/ActiveRecordTest.php]: I'll split this method

Comment (xepozz) [reply tests/framework/db/ActiveRecordTest.php]: This method contains eight test cases

**Review by xepozz — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

File: framework/db/ActiveQuery.php [hunk 4]
```diff
@@ -486,6 +486,16 @@ private function buildJoinWith()
         }
         $this->join = array_values($uniqueJoins);
 
+        // https://github.com/yiisoft/yii2/issues/16092
+        $uniqueJoinsByTableName = [];
```
Comment (jurchiks) [inline framework/db/ActiveQuery.php]: Does this work for aliased tables or only table names?
It should work primarily on aliases, and only use table name if an alias is not present.

Comment (germanow) [reply framework/db/ActiveQuery.php]: It works for both aliased tables and table names.

**Pull request merged**